### PR TITLE
Implied volatility calculation based on lets be rational paper

### PR DESCRIPTION
**Pull request by lukaskiss222**
This implements the **implied_volatility** function to calculate IV for the `black-scholes` model, based on issue #154.

## erfcx
We have added a new dependency called `errorfunctions`, which implements the `erfcx` function. The issue is that implementing it as `erfc(x) * exp(x^2)` leads to numerical issues. Therefore, we have utilized the `errorfunctions` module, which is much faster and maintains the same accuracy as the [calerf](http://www.netlib.org/specfun/erf) rational-Chebyshev code by W. J. Cody. (This is used by the default implementation of 'Let's Be Rational'). Unfortunately, the `statrs` crate does not implement `erfcx`, so we have opted for `errorfunctions`.

## Precision
For real-world cases, this method results in near `EPSILON` precision. The issue is the conversion from normalized price to actual price in extreme cases.

## Code changes outside implied volatility calculation
We have modified `year_fraction` to be public, so we can easily retrieve the `T` value for testing. It can also be useful for other users of the BSM model.

### CODE REVIEW

**Comment by lukaskiss222** [comment]
By conducting additional testing, the precision of the implementation can be enhanced. This is achievable by utilizing the `erfc` function from the `errorfunctions` library, rather than `erf::erfc` from `statrs`, for implementing the `cdf` of a Gaussian distribution.

```rust
// using erf::erfc
let normal = Gaussian::new(0.0, 1.0);
let s : f64 = -2.59949214248298;
dbg!(normal.cdf(s)); //0.00466809079256102
// using errorfunctions
dbg!(normal.cdf(s)); //0.004668090792758052
// Correct value is: 0.0046680907927580515903461813
```

I would suggest to @avhz considering a switch to `errorfunctions`, or we could explore implementing the approach used in the `scipy` [implementation.](https://github.com/scipy/scipy/blob/ae026f72fbc3b6c60a5d5b9d2e392b5c1919fd44/scipy/special/cephes/fcdf.c#L201)

Making this change will enhance accuracy by a factor of 10,000 in the `lets_be_rational`.

**Comment by avhz** [comment]
Hey mate, looks good.
Currently waiting on a new laptop so I can't play around with it yet. 
It's your call regarding the choice of error function to be used.

**Comment by lukaskiss222** [comment]
Updating the `cdf` function for Gaussian distributions resulted in the test `test_gaussian_distribution_function` not passing. After rechecking the test cases with `mpmath` and `sympy` (libraries that can calculate values with any specified precision), it appears that the values were incorrect. This issue has been fixed, and now the tests are passing again.

However, the following tests are still not passing:
- `black_scholes_1973`
- `merton_1973`

After extensive testing, I realized that the issue stems from float precision. Both Rust and standard Python cannot achieve better precision than `2.10445589535083`.

It is likely that the same issue affects the second test as well.
After we fix these two tests, the PR is ready to be pushed.

File: src/statistics/distributions/gaussian.rs [hunk 1]
```diff
@@ -351,15 +353,17 @@ mod tests_gaussian {
         let dist: Gaussian = Gaussian::default();
 
         // Values from WolframAlpha
-        assert_approx_equal!(dist.cdf(-4.0), 0.000_031_671_241_835_663_76, EPS);
-        assert_approx_equal!(dist.cdf(-3.0), 0.001_349_898_031_574_464_2, EPS);
-        assert_approx_equal!(dist.cdf(-2.0), 0.022_750_131_947_162_62, EPS);
-        assert_approx_equal!(dist.cdf(-1.0), 0.158_655_253_945_057_25, EPS);
+        assert_approx_equal!(dist.cdf(-5.0), 2.866_515_718_791_939e-7, EPS);
```
Comment (avhz) [inline src/statistics/distributions/gaussian.rs:356]: Why have these test values changed ? I pulled them from Wolfram Alpha.

Comment (lukaskiss222) [reply src/statistics/distributions/gaussian.rs:356]: Hi, 
the test was not passing and the algorithm seemed to be correct. So I rechecked the values with
- `Wolfram Alpha` using `CDF[NormalDistribution[0, 1], x] to 17 digits` where x is the parameter
- `mpmath` using `mp.ncdf` and `mp.dps=20`

Both resulted in the same numbers, but the number was different from the previous ones.
I think that is the result of changing
```rust
       // from
        0.5 * erf::erfc(-(x - self.mean) / (SQRT_2 * self.variance.sqrt()));
       // to
        0.5 * (-(x - self.mean) / (SQRT_2 * self.variance.sqrt())).erfc();
        
```

File: src/instruments/options/implied_volatility/lets_be_rational/mod.rs [hunk 2]
```diff
@@ -0,0 +1,852 @@
...
+const SQRT_TWO_PI: f64 = 2.506_628_274_631_000_7;
+
+const SQRT_DBL_MIN: f64 = 1.491_668_146_240_041_3e-154;
+const SQRT_DBL_MAX: f64 = 1.340_780_792_994_259_6e154;
+// Set this to 0 if you want positive results for (positive) denormalized inputs, else to DBL_MIN.
+// Note that you cannot achieve full machine accuracy from denormalized inputs!
+const DENORMALIZATION_CUTOFF: f64 = 0.0;
+
+const IMPLIED_VOLATILITY_MAXIMUM_ITERATIONS: usize = 2;
+const ASYMPTOTIC_EXPANSION_ACCURACY_TRERESHOLD: f64 = -10.0;
```
[the remark this answers is not in the document]
Comment (lukaskiss222) [reply src/instruments/options/implied_volatility/lets_be_rational/mod.rs]: Should be fixed for now.

**Review by avhz — APPROVED** [review]

**Pull request merged**